File: 2022/Day10/sln.hs
main :: IO ()
main = do 
    input <- readFile "Day10/input.txt"
    let all = findAccValues $ lines input
    let onesInScope =  filterCorrectVals all 
    print $ sum $ mapToSignalStrengths onesInScope

mapToSignalStrengths :: [(Int, Int)] -> [Int]
mapToSignalStrengths = map $ uncurry (*)

filterCorrectVals :: [(Int, Int)]-> [(Int, Int)]
filterCorrectVals = filter (\(_, cycle) -> cycle `elem` cycles)

cycles :: [Int]
cycles = take 6 [20 + 40*i | i <- [0..]]

findAccValues :: [String] -> [(Int, Int)]
findAccValues xs = concat $ scanl foldFn [(1,1)] xs

foldFn:: [(Int, Int)] -> String -> [(Int, Int)]
foldFn xs s 
    | "addx" `elem` s' = [(x, cycle + 1), (x + (read (last s') :: Int ), cycle + 2)]
    | "noop" `elem` s' = [(x, cycle + 1)]
    | otherwise = [(x, cycle)]
    where s' = words s
          (x, cycle) = last xs

File: 2022/Day10/sln.py
with open("2022/Day10/test.txt") as f:
    inp = f.read().splitlines()

print([20+(40*i) for i in range(6)])

acc = 0
prev = [0,0]
cycle = 0

for x in inp:
    cycle+=1
    if cycle in [20+(40*i) for i in range(6)]:
        print(acc)
    acc+=prev.pop(0)
    if x.startswith("noop"):
        prev.append(0)
    else:
        prev.append(int(x[5:]))
print(cycle)
